### ISSUE TITLE
Implement the `vanillafullstrong` expert

### ISSUE THREAD

**Issue by gasse** [ds4dm/ecole#50]
Can be implemented as an ObservationFunction in ECOLE.
Will need SCIP 7.0

**Comment by gasse** [comment]
Relevant C functions:


The branching rule has to be executed first in order for the SB scores to be computed. This can be done similarly to what is done here:
https://github.com/ds4dm/PySCIPOpt/blob/ml-branching/src/pyscipopt/scip.pyx#L5031

Basically, use [SCIPfindBranchrule](https://scip.zib.de/doc/html/group__PublicBranchRuleMethods.php#gac1d36f7900b999ecf91725b2abf30a13) to get the `vanillafullstrong` branching rule, then call [SCIPbranchruleExecLPSol](https://scip.zib.de/doc/html/branch_8c.php#a33e0fd76bd73d92527b49645632080b9) to execute it.

Finally, the SB scores can be recovered by calling [SCIPgetVanillafullstrongData](https://scip.zib.de/doc/html/branch__vanillafullstrong_8c.php#a2548062bc0760f6f06259400768ca735).

Note that the branching rule must be parameterized in a specific way to compute and collect SB scores only, without any side-effect. Those parameters are:
`branching/vanillafullstrong/integralcands` = True
`branching/vanillafullstrong/scoreall` = True
`branching/vanillafullstrong/collectscores` = True
`branching/vanillafullstrong/donotbranch` = True
`branching/vanillafullstrong/idempotent` = True

I think a good way to hack our way through this is to backup the value of those parameters first, then set them to the desired values, then call `SCIPfindBranchrule`, `SCIPbranchruleExecLPSol` and `SCIPgetVanillafullstrongData` to get the SB scores, and finally restore the parameters to the backup value. The whole process should act as an idempotent function, and leave SCIP in the same state before / after the observationfunction is called (as much as possible).

**Comment by dchetelat** [comment]
I checked the code. Right now we have e.g.
```
> obs, action_set, done = env.reset("instance_1.lp")
> obs.candidate_scores.shape
(171, 1)
> obs.candidates.shape
(171, 1)
> len(action_set)
101
```
because there are 171 pseudo-branching candidates, but I made the choice of using regular branching candidates, and there's 101 of them. It's ok, but I feel it gives a slight slant towards usage of pseudo-branching candidates. Also, having a `obs.candidates` and `obs.nb_candidates` seems redundant with the `action_set`. Instead, I would propose the following:

1. `obs.scores` should be an `np.array` of size `nb_variables` (1000 in this case), where variables that don't have a score are simply `np.nan`. (We already have `np.nan`'s in the NodeBipartiteObs so that's not weird.)
2. You can keep `obs.best_candidate` if you want but make it with respect to all variables, not the pseudo-branching candidates, that is, the new `best_candidate` should be the current `obs.candidates[obs.best_candidate]`. Maybe rename it `best_variable`?
3. Remove `obs.candidates` and `obs.nb_candidates`.

**Comment by dchetelat** [comment]
Also, your pseudo-candidates don't seem to be in order...
```
> obs.candidates.reshape(-1).astype(int)
array([154, 166, 135, 130,  25,  97,  92, 133, 131,  89,  96, 136,  38,
       115, 107,  70,  37, 167, 106, 145, 157, 122,  90,  93,  64, 105,
        86,  62,  67,  49, 160, 112,  17,  15, 141, 146,  19,  83,  29,
        52,  56, 163,  27, 151, 170, 140,  57,  39,   2,  18,  65,  58,
        41, 159, 123, 143, 116,  99,  73,  20, 139, 114,  95,  23,  35,
        36, 109, 147, 101, 148,  51,  72, 161, 104, 162, 103, 118, 126,
       129,  69,  32,  60,  33,   3, 134,  47,  31,  44,   9,  63,   6,
        30,  13,  61,  10,  76,  11, 138,  98, 137,  94,  42, 117,   4,
        28,  22, 158, 169, 119, 110,   8,  77, 108,  75,  24, 149,  88,
       168,  43, 156,  91,  14,  78, 153,  53,   5,  46,  45,  79, 124,
        26,  16,  59, 150,  74, 152, 113,  87,  85,  84,  81,  80,  48,
        34, 132,  40,  71, 125,  12,   1,  54, 128, 144, 120,  66, 102,
       111,   0, 100, 127,   7, 165, 155,  21,  50,  55,  82,  68, 121,
       164, 142])
```
I would be careful about that!

**Comment by jdumouchelle** [comment]
I agree, this is definitely a simpler representation.  I initially just went with how SCIP returned values.  I will look into the edit you just made regarding the permutation and change the representation you described above.

**Comment by dchetelat** [comment]
Okay after debugging a bit I realized that the obs.candidates were not in order because getVanillastrongbranchingData does not return the candidates in order either. Why is mysterious to me, but it doesn't matter.

**Comment by gasse** [comment]
Looks like this guy here is responsible:
https://scip.zib.de/doc/html/branch_8c_source.php#l00995

**Comment by dchetelat** [comment]
> Looks like this guy here is responsible:
> https://scip.zib.de/doc/html/branch_8c_source.php#l00995

Interesting! Anyways it doesn't matter anymore with the new system.
